pkg/admission/validator/seed.go

```diff
@@ -0,0 +1,121 @@
+// seedValidator validates create and update operations on Seed resources,
+// enforcing immutability of backup configurations.
+type seedValidator struct {
+	client         client.Client
+	decoder        runtime.Decoder
+	lenientDecoder runtime.Decoder
+}
+
+// Validate validates the Seed resource during create or update operations.
+// It enforces immutability policies on backup configurations to prevent
+// disabling immutable settings, reducing retention periods, or changing retention types.
+func (s *seedValidator) Validate(ctx context.Context, newObj, oldObj client.Object) error {
+	newSeed, ok := newObj.(*core.Seed)
+	if !ok {
+		return fmt.Errorf("wrong object type %T for new object", newObj)
+	}
+
+	if oldObj != nil {
+		oldSeed, ok := oldObj.(*core.Seed)
+		if !ok {
+			return fmt.Errorf("wrong object type %T for old object", oldObj)
+		}
+		return s.validateUpdate(ctx, oldSeed, newSeed)
+	}
+
+	return s.validateCreate(newSeed)
+}
+
+// validateCreate validates the Seed object upon creation.
+// It checks if immutable settings are provided and validates them to ensure they meet the required criteria.
+func (s *seedValidator) validateCreate(newSeed *core.Seed) error {
+	backupBucketConfig, err := admission.DecodeSeedBackupBucketConfig(s.decoder, newSeed.Spec.Backup)
+	if err != nil {
+		return fmt.Errorf("error decoding BackupBucketConfig: %w", err)
+	}
+
+	allErrs := gcpvalidation.ValidateBackupBucketConfig(backupBucketConfig, field.NewPath("spec", "backup", "providerConfig"))
+	if len(allErrs) > 0 {
+		return fmt.Errorf("validation failed: %w", allErrs.ToAggregate())
+	}
+
+	return nil
+}
+
+// validateUpdate validates updates to the Seed resource, ensuring that immutability settings for backup buckets
+// are correctly managed. It enforces constraints such as preventing the unlocking of retention policies,
+// disabling immutability once locked, and reduction of retention periods when policies are locked.
+func (s *seedValidator) validateUpdate(_ context.Context, oldSeed, newSeed *core.Seed) error {
+	oldBackupBucketConfig, err := admission.DecodeSeedBackupBucketConfig(s.lenientDecoder, oldSeed.Spec.Backup)
+	if err != nil {
+		return fmt.Errorf("error decoding old BackupBucketConfig: %w", err)
+	}
+
+	if oldBackupBucketConfig == nil || oldBackupBucketConfig.Immutability == (gcp.ImmutableConfig{}) {
+		return s.validateCreate(newSeed)
+	}
+
+	newBackupBucketConfig, err := admission.DecodeSeedBackupBucketConfig(s.decoder, newSeed.Spec.Backup)
+	if err != nil {
+		return fmt.Errorf("error decoding new BackupBucketConfig: %w", err)
+	}
+
+	allErrs := gcpvalidation.ValidateBackupBucketConfig(newBackupBucketConfig, field.NewPath("spec", "backup", "providerConfig"))
+	if len(allErrs) > 0 {
+		return fmt.Errorf("validation failed: %w", allErrs.ToAggregate())
+	}
+
+	if !oldBackupBucketConfig.Immutability.Locked {
+		return nil
+	}
+
+	if newBackupBucketConfig == nil {
+		return fmt.Errorf("immutability cannot be disabled once it is locked")
```

Comment:
I don't think we should rely on nil checks for the config. It is better to treat the a nil value in the providerConfig as the "zero" value `BackupBucketConfig`.

Reason: the moment that another field aside from immutability is introduced in the Config, these checks need to be re-written.

Reply:
Now I introduced new function `validateImmutabilityUpdate` to validate the immutability on update, In future, we can introduce new function to validate for another fields
Fixed as part of c12aa8c1415b48b0a4fb0a478dc24c604c6523a1